### ISSUE TITLE
Storage tests for COO

### ISSUE THREAD

**Issue by KyleVaughn** [bkochuna/ners570f24-SpMV#51]
## Definition of done:

**Comment by xianyu2023umich** [comment]
I'm reading it. Very comprehensive definition of done.

**Comment by xianyu2023umich** [comment]
@athenafhl Hey I'm reviewing you test it fails the compile at line 34: SpMV::SparseMatrix<T>* ptr_A = new SpMV::SparseMatrix_COO<T>(n, n); it's probably cuz you declared fp_type in the template instead of T.

**Comment by athenafhl** [comment]
> @athenafhl Hey I'm reviewing you test it fails the compile at line 34: SpMV::SparseMatrix* ptr_A = new SpMV::SparseMatrix_COO(n, n); it's probably cuz you declared fp_type in the template instead of T.

Hi! I fix all the bugs! I think the rest of the errors come from the matvec test, not the storage test

**Comment by xianyu2023umich** [comment]
> > @athenafhl Hey I'm reviewing you test it fails the compile at line 34: SpMV::SparseMatrix* ptr_A = new SpMV::SparseMatrix_COO(n, n); it's probably cuz you declared fp_type in the template instead of T.
> 
> Hi! I fix all the bugs! I think the rest of the errors come from the matvec test, not the storage test. I can do nothing until the other people update their branches.

Tried your test again passed the compiler but the test failed:

[ctest]    Site: gl-login3.arc-ts.umich.edu
[ctest]    Build name: Linux-g++
[ctest] Test project /home/xianyu/hw3/ners570f24-SpMV/build
[ctest]     Start 1: test_COO_storage_test
[ctest] 1/1 Test #1: test_COO_storage_test ............Subprocess aborted***Exception:   5.43 sec
[ctest] test_COO_storage_test: /home/xianyu/hw3/ners570f24-SpMV/tests/COO_storage_test.cpp:58: void random_nxn() [with long unsigned int n = 100; fp_type = double]: Assertion `I[i] == rowIdx[i]' failed.
[ctest] 
[ctest] 
[ctest] 0% tests passed, 1 tests failed out of 1
[ctest] 
[ctest] Total Test time (real) =   5.45 sec
[ctest] 
[ctest] The following tests FAILED:
[ctest] 	  1 - test_COO_storage_test (Subprocess aborted)
[ctest] Errors while running CTest

Is this due to the matvec?

**Comment by xianyu2023umich** [comment]
> > > @athenafhl Hey I'm reviewing you test it fails the compile at line 34: SpMV::SparseMatrix* ptr_A = new SpMV::SparseMatrix_COO(n, n); it's probably cuz you declared fp_type in the template instead of T.
> > 
> > 
> > Hi! I fix all the bugs! I think the rest of the errors come from the matvec test, not the storage test. I can do nothing until the other people update their branches.
> 
> Tried your test again passed the compiler but the test failed:
> 
> [ctest] Site: gl-login3.arc-ts.umich.edu [ctest] Build name: Linux-g++ [ctest] Test project /home/xianyu/hw3/ners570f24-SpMV/build [ctest] Start 1: test_COO_storage_test [ctest] 1/1 Test #1: test_COO_storage_test ............Subprocess aborted***Exception: 5.43 sec [ctest] test_COO_storage_test: /home/xianyu/hw3/ners570f24-SpMV/tests/COO_storage_test.cpp:58: void random_nxn() [with long unsigned int n = 100; fp_type = double]: Assertion `I[i] == rowIdx[i]' failed. [ctest] [ctest] [ctest] 0% tests passed, 1 tests failed out of 1 [ctest] [ctest] Total Test time (real) = 5.45 sec [ctest] [ctest] The following tests FAILED: [ctest] 1 - test_COO_storage_test (Subprocess aborted) [ctest] Errors while running CTest
> 
> Is this due to the matvec?

Oh no I think it's due to the storage has nothing to do with matvec.

**Comment by athenafhl** [comment]
> > > > @athenafhl Hey I'm reviewing you test it fails the compile at line 34: SpMV::SparseMatrix* ptr_A = new SpMV::SparseMatrix_COO(n, n); it's probably cuz you declared fp_type in the template instead of T.
> > > 
> > > 
> > > Hi! I fix all the bugs! I think the rest of the errors come from the matvec test, not the storage test. I can do nothing until the other people update their branches.
> > 
> > 
> > Tried your test again passed the compiler but the test failed:
> > [ctest] Site: gl-login3.arc-ts.umich.edu [ctest] Build name: Linux-g++ [ctest] Test project /home/xianyu/hw3/ners570f24-SpMV/build [ctest] Start 1: test_COO_storage_test [ctest] 1/1 Test #1: test_COO_storage_test ............Subprocess aborted***Exception: 5.43 sec [ctest] test_COO_storage_test: /home/xianyu/hw3/ners570f24-SpMV/tests/COO_storage_test.cpp:58: void random_nxn() [with long unsigned int n = 100; fp_type = double]: Assertion `I[i] == rowIdx[i]' failed. [ctest] [ctest] [ctest] 0% tests passed, 1 tests failed out of 1 [ctest] [ctest] Total Test time (real) = 5.45 sec [ctest] [ctest] The following tests FAILED: [ctest] 1 - test_COO_storage_test (Subprocess aborted) [ctest] Errors while running CTest
> > Is this due to the matvec?
> 
> Oh no I think it's due to the storage has nothing to do with matvec.

Yeah you are right. I delete the matvec part and the test did not pass. Did you see the assemble method for COO? Would it be the problem of creating the storage? It looks pretty weird.

**Comment by xianyu2023umich** [comment]
> > > > > @athenafhl Hey I'm reviewing you test it fails the compile at line 34: SpMV::SparseMatrix* ptr_A = new SpMV::SparseMatrix_COO(n, n); it's probably cuz you declared fp_type in the template instead of T.
> > > > 
> > > > 
> > > > Hi! I fix all the bugs! I think the rest of the errors come from the matvec test, not the storage test. I can do nothing until the other people update their branches.
> > > 
> > > 
> > > Tried your test again passed the compiler but the test failed:
> > > [ctest] Site: gl-login3.arc-ts.umich.edu [ctest] Build name: Linux-g++ [ctest] Test project /home/xianyu/hw3/ners570f24-SpMV/build [ctest] Start 1: test_COO_storage_test [ctest] 1/1 Test #1: test_COO_storage_test ............Subprocess aborted***Exception: 5.43 sec [ctest] test_COO_storage_test: /home/xianyu/hw3/ners570f24-SpMV/tests/COO_storage_test.cpp:58: void random_nxn() [with long unsigned int n = 100; fp_type = double]: Assertion `I[i] == rowIdx[i]' failed. [ctest] [ctest] [ctest] 0% tests passed, 1 tests failed out of 1 [ctest] [ctest] Total Test time (real) = 5.45 sec [ctest] [ctest] The following tests FAILED: [ctest] 1 - test_COO_storage_test (Subprocess aborted) [ctest] Errors while running CTest
> > > Is this due to the matvec?
> > 
> > 
> > Oh no I think it's due to the storage has nothing to do with matvec.
> 
> Yeah you are right. I delete the matvec part and the test did not pass. Did you see the assemble method for COO? Would it be the problem of creating the storage? It looks pretty weird.

yeah I don't really understand that loop (especially the head) but definitely it shouldn't be index += index should be +=1

**Comment by athenafhl** [comment]
> > > > > > @athenafhl Hey I'm reviewing you test it fails the compile at line 34: SpMV::SparseMatrix* ptr_A = new SpMV::SparseMatrix_COO(n, n); it's probably cuz you declared fp_type in the template instead of T.
> > > > > 
> > > > > 
> > > > > Hi! I fix all the bugs! I think the rest of the errors come from the matvec test, not the storage test. I can do nothing until the other people update their branches.
> > > > 
> > > > 
> > > > Tried your test again passed the compiler but the test failed:
> > > > [ctest] Site: gl-login3.arc-ts.umich.edu [ctest] Build name: Linux-g++ [ctest] Test project /home/xianyu/hw3/ners570f24-SpMV/build [ctest] Start 1: test_COO_storage_test [ctest] 1/1 Test #1: test_COO_storage_test ............Subprocess aborted***Exception: 5.43 sec [ctest] test_COO_storage_test: /home/xianyu/hw3/ners570f24-SpMV/tests/COO_storage_test.cpp:58: void random_nxn() [with long unsigned int n = 100; fp_type = double]: Assertion `I[i] == rowIdx[i]' failed. [ctest] [ctest] [ctest] 0% tests passed, 1 tests failed out of 1 [ctest] [ctest] Total Test time (real) = 5.45 sec [ctest] [ctest] The following tests FAILED: [ctest] 1 - test_COO_storage_test (Subprocess aborted) [ctest] Errors while running CTest
> > > > Is this due to the matvec?
> > > 
> > > 
> > > Oh no I think it's due to the storage has nothing to do with matvec.
> > 
> > 
> > Yeah you are right. I delete the matvec part and the test did not pass. Did you see the assemble method for COO? Would it be the problem of creating the storage? It looks pretty weird.
> 
> yeah I don't really understand that loop (especially the head) but definitely it shouldn't be index += index should be +=1

Ok I'll tell him to debug that. Thank you for your help!
